openpgp/packet/encrypted_key.go

```diff
@@ -184,41 +309,106 @@ func (e *EncryptedKey) Serialize(w io.Writer) error {
 // SerializeEncryptedKey serializes an encrypted key packet to w that contains
 // key, encrypted to pub.
 // If config is nil, sensible defaults will be used.
-func SerializeEncryptedKey(w io.Writer, pub *PublicKey, cipherFunc CipherFunction, key []byte, config *Config) error {
-	var buf [10]byte
-	buf[0] = encryptedKeyVersion
-	binary.BigEndian.PutUint64(buf[1:9], pub.KeyId)
-	buf[9] = byte(pub.PubKeyAlgo)
-
-	keyBlock := make([]byte, 1 /* cipher type */ +len(key)+2 /* checksum */)
-	keyBlock[0] = byte(cipherFunc)
-	copy(keyBlock[1:], key)
-	checksum := checksumKeyMaterial(key)
-	keyBlock[1+len(key)] = byte(checksum >> 8)
-	keyBlock[1+len(key)+1] = byte(checksum)
+func SerializeEncryptedKey(w io.Writer, pub *PublicKey, cipherFunc CipherFunction, aeadSupported bool, key []byte, config *Config) error {
+	var buf [35]byte // max possible header size is v6
+	lenHeaderWritten := 1
+	version := 3
+
+	if aeadSupported {
+		version = 6
+	}
+	// An implementation MUST NOT generate ElGamal v6 PKESKs.
+	if version == 6 && pub.PubKeyAlgo == PubKeyAlgoElGamal {
+		return errors.InvalidArgumentError("ElGamal v6 PKESK are not allowed")
+	}
+	// In v3 PKESKs, for X25519 and X448, mandate using AES
+	if version == 3 && (pub.PubKeyAlgo == PubKeyAlgoX25519 || pub.PubKeyAlgo == PubKeyAlgoX448) {
+		switch cipherFunc {
+		case CipherAES128, CipherAES192, CipherAES256:
+			break
+		default:
+			return errors.InvalidArgumentError("v3 PKESK mandates AES for x25519 and x448")
+		}
+	}
+
+	buf[0] = byte(version)
+
+	if version == 6 {
+		if pub != nil {
+			buf[1] = byte(pub.Version)
+			copy(buf[2: len(pub.Fingerprint)+2], pub.Fingerprint)
+			lenHeaderWritten += len(pub.Fingerprint) + 1
+		} else {
+			// anonymous case
+			buf[1] = 0
+			lenHeaderWritten += 1
+		}
+	} else {
+		binary.BigEndian.PutUint64(buf[1:9], pub.KeyId)
+		lenHeaderWritten += 8
+	}
+	buf[lenHeaderWritten] = byte(pub.PubKeyAlgo)
+	lenHeaderWritten += 1
+
+	var keyBlock []byte
+	// for X25519 there is no checksum and a special encoding
+	if pub.PubKeyAlgo == PubKeyAlgoX25519 || pub.PubKeyAlgo == PubKeyAlgoX448 {
+		// no checksum is added
+		if version < 6 {
+			// For v3 PKESK packets, seven zero-octets are 
+			// added as padding after the algorithm identifier
+			keyBlock = make([]byte, len(key) + 8)
+			keyBlock[0] = byte(cipherFunc)
+			copy(keyBlock[8:], key)
```

Comment:
This was changed in https://gitlab.com/openpgp-wg/rfc4880bis/-/merge_requests/266